ge/org.osate.ge.ba/src/org/osate/ge/ba/ui/properties/BehaviorTransitionPropertySection.java

```diff
@@ -95,56 +101,199 @@ public void createControls(final Composite parent, final TabbedPropertySheetPage
+			final String btnId) {
+		editingControls.refresh();
+		editingControls.setStyledTextTestId(styledTextId);
+		editingControls.setButtonTestId(btnId);
+	}
+
+	private static SelectionAdapter getEditActionSelectionAdapter(final IProject project,
+			final EmbeddedTextValue actionTextValue, final BehaviorTransition behaviorTransition,
+			final TransactionalEditingDomain editingDomain, final IXtextDocument xtextDocument,
+			final XtextResource xtextResource, final String src) {
+		return new SelectionAdapter() {
+			@Override
+			public void widgetSelected(final SelectionEvent e) {
+				final EmbeddedXtextAdapter xtextAdapter = new EmbeddedXtextAdapter(project, actionTextValue);
+				final BiFunction<EObject, String, String> getModifiedSrc = (rootElement, newActionBlock) -> {
+					String modifiedSrc = xtextAdapter.serialize(rootElement);
+					if (newActionBlock.isEmpty()) {
+						// Remove brackets for empty action block
+						final String prefix = actionTextValue.getPrefix();
+						modifiedSrc = prefix.substring(0, prefix.length() - 1)
+								+ actionTextValue.getSuffix().substring(1);
+					}
+
+					return modifiedSrc;
+				};
+
+				final BiFunction<BehaviorTransition, String, Boolean> isValidModification = (bt, newText) -> {
+					final BehaviorActionBlock actionBlock = bt.getActionBlock();
+					// Calculate enabled based on if action should exist and if it exists
+					return newText.isEmpty() ? actionBlock == null : actionBlock != null;
+				};
+
+				final EditEmbeddedTextDialog dlg = new EditEmbeddedTextDialog(Display.getCurrent().getActiveShell(),
+						"Edit Transition Action", "Enter new action block.", xtextAdapter,
+						SWT.BORDER | SWT.V_SCROLL | SWT.WRAP | SWT.MULTI,
+						GridDataFactory.fillDefaults().indent(10, 0).grab(true, true).hint(SWT.DEFAULT, 100).create(),
+						behaviorTransition,
+						getModifiedSrc, isValidModification);
+				if (dlg.open() == Window.OK) {
+					// Edit action
+					BehaviorAnnexSelectionUtil.getActiveEditor().ifPresent(editorPart -> {
+						final ActionService actionService = Adapters.adapt(editorPart, ActionService.class);
+						final ModelChangeNotifier modelChangeNotifier = Objects.requireNonNull(
+								editorPart.getAdapter(ModelChangeNotifier.class),
+								"Unable to get model change notifier");
+
+						if (xtextDocument != null) {
+							// Get source text for xtext document
+							final String srcText = dlg.getText(true);
+							// Execute modification with xtext document
+							actionService.execute("Modifying Behavior Transition Action Block", ExecutionMode.NORMAL,
+									new EmbeddedTextModificationAction(xtextDocument, modelChangeNotifier, project,
+											srcText));
+						} else {
+							// Execute modification with xtext resource
+							final boolean actionExists = !actionTextValue.getEditableText().isEmpty();
+							final String newText = getNewText(dlg.getText(false), actionExists);
```

Comment:
getText(boolean) is called with false here for editing the XtextResource (editor closed).

Reply:
Okay, I missed that . if that is need then that's fine. In general I was confused a bit about all the modified text stuff. It wasn't clear what some of that was doing.